src/Raven.Pal/src/win/pager.c

```diff
@@ -558,3 +696,176 @@ int32_t rvn_pager_get_file_handle(
     *file_handle = h;
     return SUCCESS;
 }
+
+int32_t rvn_write_file_io(
+    void* handle,
+    int32_t count,
+    struct page_to_write *buffers,
+    int32_t *detailed_error_code
+)
+{
+    struct handle *handle_ptr = handle;
+    OVERLAPPED ov = {0};
+    for(int32_t i = 0; i < count; i++)
+    {
+        int64_t offset = buffers[i].page_num * VORON_PAGE_SIZE;
+        int64_t size = (int64_t)buffers[i].count_of_pages * VORON_PAGE_SIZE;
+        while(size > 0)
+        {
+            ov.Offset = (DWORD)offset;
+            ov.OffsetHigh = (DWORD)(offset >> 32);
+
+            int32_t size_to_write = (int32_t)(rvn_min(size, INT32_MAX));
+            if(!WriteFile(handle_ptr->file_handle, buffers[i].ptr, size_to_write, NULL, &ov))
+            {
+                *detailed_error_code = GetLastError();
+                return FAIL_WRITE_FILE;
+            }    
+            offset += size_to_write;
+            size -= size_to_write;
+        }
+    }
+    return SUCCESS;
+}
+
+int32_t _submit_and_wait(
+    HIORING io_ring,
+    int32_t count,
+    int32_t* detailed_error_code)
+{
+    HRESULT hr = SubmitIoRing(io_ring, count, INFINITE, NULL);
+    if (FAILED(hr))
+    {
+        *detailed_error_code = hr;
+        return FAIL_IO_RING_SUBMIT;
+    }
+    IORING_CQE cqe;
+    for(int i = 0; i < count; i++)
+    {
+        hr = PopIoRingCompletion(io_ring, &cqe);
+        if (hr != S_OK)
+        {
+            *detailed_error_code = hr;
+            return FAIL_IO_RING_NO_RESULT;
+        }
+        if(FAILED(cqe.ResultCode))
+        {
+            *detailed_error_code = cqe.ResultCode;
+            return FAIL_IO_RING_WRITE_RESULT;
+        }
+    }
+    return SUCCESS;
+}
+
+int32_t rvn_write_io_ring(
+    void* handle,
+    int32_t count,
+    struct page_to_write *buffers,
+    int32_t *detailed_error_code
+)
+{
+    struct handle *handle_ptr = handle;
+    IORING_HANDLE_REF file_handle_ref =  IoRingHandleRefFromHandle(handle_ptr->file_handle);
+    uint32_t submitted = 0;
+    HRESULT hr;
+    int32_t rc = SUCCESS;
+
+    EnterCriticalSection(&handle_ptr->global_state->lock);
+    for(int32_t i = 0; i < count; i++)
+    {
+        IORING_BUFFER_REF buf = IoRingBufferRefFromPointer(buffers[i].ptr);
+        int64_t offset = buffers[i].page_num * VORON_PAGE_SIZE;
+        int64_t size = (int64_t)buffers[i].count_of_pages * VORON_PAGE_SIZE;
+        while(size > 0)
+        {
+            int32_t size_to_write = (int32_t)(rvn_min(size, INT32_MAX));
+            hr = BuildIoRingWriteFile(handle_ptr->global_state->io_ring, file_handle_ref, 
+                buf, size_to_write, offset, FILE_WRITE_FLAGS_NONE, 0,
+                IOSQE_FLAGS_NONE);
+            size -= size_to_write;
+            offset += size_to_write;
+
+            if (FAILED(hr))
+            {
+                *detailed_error_code = hr;
+                return FAIL_IO_RING_WRITE;
```

Comment:
I am not that well versed into `EnterCriticalSection` but it seems that if write fails, we are not going out from it.